test/test_torch.py

```diff
@@ -6258,6 +6258,12 @@ def test_scalar_check(self, device):
         self.assertEqual((), torch.normal(zero_d, 1).shape)
         self.assertEqual((1,), torch.normal(one_d, 1).shape)
 
+        # convolutions.  Yes, we are testing nn.functional here; seems justified
+        # given its similar to the other tests
+        w = torch.randn(2, 1, 3, 3, device=device).div_(2).requires_grad_()
+        self.assertRaises(RuntimeError, lambda: torch.nn.functional.conv2d(zero_d, w, groups=1))
```

Comment:
This will always be routed to the `thnn` version? What about the mklnn if it's available?

Reply:
MKLDNN is only on CPU.

For CUDA it depends on cudnn is available or not -- I'm not sure if we have any tests of that case but it doesn't _really_ matter -- what matters is that the function handles scalars correctly.

Reply:
Is this test CPU or GPU?

In any case, since we have different backends that all need to handle this case properly, we should test all the backends no?

Reply:
I don't know what backend means, but this test does check all the device types.

Reply:
My understanding is that now that the codegen does not take care of this but the low level implementation does. Every single possible low level implementation must handle this properly.

As an example:
If we run this on GPU. either cudnn or thcunn will be checked. But not both. Right?